### PR TITLE
 Fix #2793 rework quic ack creator

### LINKED ISSUE #2793: The _packet_numbers will get negative value
The _packet_numbers will get negative value  when we recv a small packet after do large ack.

ping @masaori335 @maskit 

```
TEST_CASE("QUICAckFrameCreator_last_ack_number", "[quic]")
{
  QUICAckFrameCreator creator;
  std::unique_ptr<QUICAckFrame, QUICFrameDeleterFunc> frame = {nullptr, nullptr};

  creator.update(6, true);
  creator.update(7, true);
  frame = creator.create();
  CHECK(frame != nullptr);
  CHECK(frame->num_blocks() == 0);
  CHECK(frame->largest_acknowledged() == 7);
  CHECK(frame->ack_block_section()->begin()->gap() == 0);

  creator.update(4, true);
  creator.update(5, true);
}
```

```
QUICAckFrameCreator::update (this=0x7fffffffd190, packet_number=4, acknowledgable=true) at ../QUICAckFrameCreator.cc:30
30	  if (this->_packet_count == MAXIMUM_PACKET_COUNT) {
(gdb) n
33	  if (packet_number > this->_largest_ack_number) {
(gdb) 
37	  this->_packet_numbers[this->_packet_count++] = packet_number - this->_last_ack_number;
(gdb) p this->_packet_numbers
$1 = {6, 7, 0, 0, 0, 3072, 0, 0, 4912, 62977, 32767, 0 <repeats 65 times>, 41224, 63221, 32767, 0, 0, 0, 0, 0, 0, 1024, 0, 0, 4912, 62977, 32767, 0, 0, 0, 0, 0, 41312, 63221, 
  32767, 0, 45698, 78, 0, 0, 53473, 78, 0, 0, 46218, 76, 0, 0, 2741, 79, 0, 0, 2164, 79, 0, 0, 39002, 77, 0, 0, 65349, 62834, 32767, 0, 2, 0, 0, 0, 54144, 65535, 32767, 0, 
  59584, 132, 0, 0, 54144, 65535, 32767, 0, 2, 0, 0, 0, 37425, 63218, 32767, 0, 64100, 65535, 4095, 0, 54368, 65535, 32767, 0, 54048, 65535, 32767, 0, 48613, 78, 0, 0, 0, 0, 0, 
  0, 54432, 65535, 32767, 0, 4912, 62977, 32767, 0, 6, 0, 0, 0, 35507, 16821, 0, 0, 40288, 92, 0, 0, 48200, 78, 0, 0, 46544, 63216, 32767, 0, 60608, 132, 0, 0, 0, 0, 32767, 0, 
  8192, 0, 0, 0, 1701, 63455, 32767, 0, 0, 0, 0, 0, 0, 0, 0, 0, 54160, 65535, 32767, 0, 45859, 82, 0, 0, 47624, 78, 0 <repeats 22 times>}
(gdb) p this->_packet_count
$2 = 0
(gdb) n
38	  if (acknowledgable && !this->_can_send) {
(gdb) p this->_packet_count
$3 = 1
(gdb) p this->_packet_numbers
$4 = {65533, 7, 0, 0, 0, 3072, 0, 0, 4912, 62977, 32767, 0 <repeats 65 times>, 41224, 63221, 32767, 0, 0, 0, 0, 0, 0, 1024, 0, 0, 4912, 62977, 32767, 0, 0, 0, 0, 0, 41312, 
  63221, 32767, 0, 45698, 78, 0, 0, 53473, 78, 0, 0, 46218, 76, 0, 0, 2741, 79, 0, 0, 2164, 79, 0, 0, 39002, 77, 0, 0, 65349, 62834, 32767, 0, 2, 0, 0, 0, 54144, 65535, 32767, 
  0, 59584, 132, 0, 0, 54144, 65535, 32767, 0, 2, 0, 0, 0, 37425, 63218, 32767, 0, 64100, 65535, 4095, 0, 54368, 65535, 32767, 0, 54048, 65535, 32767, 0, 48613, 78, 0, 0, 0, 0, 
  0, 0, 54432, 65535, 32767, 0, 4912, 62977, 32767, 0, 6, 0, 0, 0, 35507, 16821, 0, 0, 40288, 92, 0, 0, 48200, 78, 0, 0, 46544, 63216, 32767, 0, 60608, 132, 0, 0, 0, 0, 32767, 
  0, 8192, 0, 0, 0, 1701, 63455, 32767, 0, 0, 0, 0, 0, 0, 0, 0, 0, 54160, 65535, 32767, 0, 45859, 82, 0, 0, 47624, 78, 0 <repeats 22 times>}
(gdb) n
39	    this->_can_send = true;
(gdb) q
A debugging session is active.

	Inferior 1 [process 29211] will be killed.

Quit anyway? (y or n) y
```

### PR DESCRIPTION
**Pull request by scw00**

### CODE REVIEW

File: iocore/net/quic/test/test_QUICAckFrameCreator.cc [hunk 1]
```diff
@@ -64,6 +64,80 @@ TEST_CASE("QUICAckFrameCreator", "[quic]")
...
+  creator.update(5, true);
+  creator.update(6, true);
+  creator.update(8, true);
+  creator.update(9, true);
+
+  frame = creator.create();
+  CHECK(frame != nullptr);
+  CHECK(frame->num_blocks() == 2);
+  CHECK(frame->largest_acknowledged() == 9);
   CHECK(frame->ack_block_section()->first_ack_block_length() == 2);
```
Comment (scw00) [inline iocore/net/quic/test/test_QUICAckFrameCreator.cc]: I'm not sure here, why dose the ack_block_length is 2.

Comment (maskit) [reply iocore/net/quic/test/test_QUICAckFrameCreator.cc]: Because it's for packet number 6 and 7, if I understand the spec correctly.

Comment (scw00) [reply iocore/net/quic/test/test_QUICAckFrameCreator.cc]: The spec says the first_ack_block_length is the `number of contiguous packets preceding the Largest Acknowledged`.

We only received 10 here.

```
A variable-length integer indicating the number of contiguous packets preceding the Largest Acknowledged that are being acknowledged.
```

Comment (maskit) [reply iocore/net/quic/test/test_QUICAckFrameCreator.cc]: I see.  I thought opposite (a block for 6 and 7 follows a block for 10) but probably you're right.

So, the sequence of ack blocks here has to be:
First block: for 10
Second block (the block after the first gap): for 7 and 6
Right? Then `first_ack_block_length` should be 1.

**Comment by maskit** [comment]
@scw00 By the way, are you referring the unpublished real latest spec (editor's copy)?

That one is really changing and we cannot introp with it. Let's focus on draft-07 for now. I think there are many places that are not changed between draft-07 and the editor's copy.

**Comment by scw00** [comment]
I use this spec.

https://quicwg.github.io/base-drafts/draft-ietf-quic-transport.html#rfc.section.8.16.1

**Comment by maskit** [comment]
Please refer to this one for now.
https://tools.ietf.org/id/draft-ietf-quic-transport-07.html

This list is the incompatible changes between -07 and the latest, if the list is updated.
https://quicwg.github.io/base-drafts/draft-ietf-quic-transport.html#rfc.appendix.C.1

File: iocore/net/quic/test/test_QUICAckFrameCreator.cc (hunk 1 above)
Comment (maskit) [reply iocore/net/quic/test/test_QUICAckFrameCreator.cc]: Is my understanding correct now?

I'm +1 on the change to introduce `QUICAckPacketNumbers` but I want to confirm that we are on the same.

Comment (maskit) [reply iocore/net/quic/test/test_QUICAckFrameCreator.cc]: I'm implementing ack block section parser (it's been left as TODO surprisingly) and I confirmed that your understanding is right. I'll approve this when CI jobs complete.

**Comment by maskit** [comment]
[approve ci]

**Pull request merged**